src/App/App.test.js

```diff
@@ -1,8 +1,114 @@
+import userEvent from "@testing-library/user-event";
+import "@testing-library/jest-dom";
+import { Router, MemoryRouter } from "react-router";
+import { createMemoryHistory } from "history";
+import { addToMySet } from "../actions";
+import { fetchTenJokes } from "../apiCalls";
+import mockData from "../mockData";
+import { act } from "react-dom/test-utils";
 
-test('renders learn react link', () => {
-  render(<App />);
-  const linkElement = screen.getByText(/learn react/i);
-  expect(linkElement).toBeInTheDocument();
+jest.mock("../apiCalls");
+jest.mock("../actions");
+
+describe("App", () => {
+  beforeEach(() => {
+    fetchTenJokes.mockResolvedValueOnce(mockData.jokes);
+    render(
+      <MemoryRouter>
+        <App />
+      </MemoryRouter>
+    );
+  });
+  it("should render the NavBar on load with DevWit App name", () => {
+    const nav = screen.getByRole("navigation");
+    expect(nav).toBeInTheDocument();
+  });
+
+  it("should redirect URL path to /home on load", () => {
+    const history = createMemoryHistory();
+
+    render(
+      <Router history={history}>
+        <App />
+      </Router>
+    );
+    expect(history.location.pathname).toBe("/home");
+  });
+
+  it("should render Home component on load", () => {
+    const welcomeMsg = screen.getByRole("heading", { name: /welcome, dev\./i });
+    expect(welcomeMsg).toBeInTheDocument();
+  });
+
+  it("should render GetMaterial when GetMaterial tab is clicked", () => {
+    const getMaterialTab = screen.getByRole("button", {
+      name: /get some new material/i,
+    });
+    userEvent.click(getMaterialTab);
+
+    const getMaterialHeader = screen.getByRole("heading", {
+      name: /need some new material\?/i,
+    });
+    expect(getMaterialHeader).toBeInTheDocument();
+  });
+
+  it("should render MySet component when MySet tab is clicked", () => {
+    const mySetTab = screen.getByRole("button", { name: /my set/i });
+    userEvent.click(mySetTab);
+
+    const mySetHeader = screen.getByRole("heading", {
+      name: /this is my set/i,
+    });
+    expect(mySetHeader).toBeInTheDocument();
+  });
+
+  it("should add a joke that the user likes to My Set when '+' button is clicked", async () => {
+    const funnyJoke = {
+      id: 27,
+      type: "programming",
+      setup: "['hip', 'hip']",
+      punchline: "(hip hip array)",
+    };
+
+    addToMySet.mockResolvedValueOnce(funnyJoke);
+
+    const getMaterialTab = screen.getByRole("button", {
+      name: /get some new material/i,
+    });
+
+    userEvent.click(getMaterialTab);
+
+    const getTenJokesButton = screen.getByRole("button", {
+      name: "Get 10 Jokes",
+    });
+
+    userEvent.click(getTenJokesButton);
+
+    const funnyJokeSetup = await waitFor(() =>
+      screen.getByText("['hip', 'hip']")
+    );
+
+    expect(funnyJokeSetup).toBeInTheDocument();
+
+    const addButton = screen.getByRole("button", {
+      name: /add\-joke\-27\-to\-set/i,
+    });
+
+    act(() => {
```

Comment:
Would love to hear about this act method. Haven't used this yet

Reply:
Not sure I was using it correctly, but from what I understand, it allows for updates to the UI before you make any assertions... I was attempting to use it so that when a click event happens on the 'add' button, the card is added to MySet component, and then I navigate to the MySet component to see if it's there.